orchagent/fgnhgorch.h

```diff
@@ -0,0 +1,128 @@
+#ifndef SWSS_FGNHGORCH_H
+#define SWSS_FGNHGORCH_H
+
+#include "orch.h"
+#include "observer.h"
+#include "intfsorch.h"
+#include "neighorch.h"
+#include "producerstatetable.h"
+
+#include "ipaddress.h"
+#include "ipaddresses.h"
+#include "ipprefix.h"
+#include "nexthopgroupkey.h"
+
+#include <map>
+
+typedef uint32_t Bank;
+typedef std::set<NextHopKey> ActiveNextHops;
+typedef std::vector<sai_object_id_t> FGNextHopGroupMembers;
+typedef std::vector<uint32_t> HashBuckets;
+typedef std::map<NextHopKey, HashBuckets> FGNextHopGroupMap;
+typedef std::vector<FGNextHopGroupMap> BankFGNextHopGroupMap;
+typedef std::map<Bank,Bank> InactiveBankMapsToBank;
+
+struct FGNextHopGroupEntry
+{
+    sai_object_id_t         next_hop_group_id;      // next hop group id
+    FGNextHopGroupMembers   nhopgroup_members;      // sai_object_ids of nexthopgroup members(0 - real_bucket_size - 1)
+    ActiveNextHops          active_nexthops;        // The set of nexthops(ip+alias)
+    BankFGNextHopGroupMap   syncd_fgnhg_map;        // Map of (bank) -> (nexthops) -> (index in nhopgroup_members)
+    NextHopGroupKey         nhg_key;                // Full next hop group key
+    InactiveBankMapsToBank  inactive_to_active_map; // Maps an inactive bank to an active one in terms of hash bkts
+};
+
+/*TODO: can we make an optimization here when we get multiple routes pointing to a fgnhg */
+typedef std::map<IpPrefix, FGNextHopGroupEntry> FGRouteTable;
+/* RouteTables: vrf_id, FGRouteTable */
+typedef std::map<sai_object_id_t, FGRouteTable> FGRouteTables;
+/* Name of the FG NHG group */
+typedef std::string FgNhg;
+/* Map from IP to Bank */
+typedef std::map<IpAddress, Bank> NextHops;
+/* Cache currently ongoing FG_NHG PREFIX additions/deletions */
+typedef std::map<IpPrefix, NextHopGroupKey> FgPrefixOpCache;
+
+/* Store the indices occupied by a bank */
+typedef struct
+{
+    uint32_t start_index;
+    uint32_t end_index;
+} bank_index_range;
+
+typedef struct FgNhgEntry
+{
+    string fgNhg_name;                                  // Name of FG NHG group configured by user
+    uint32_t configured_bucket_size;                    // Bucket size configured by user
+    uint32_t real_bucket_size;                          // Real bucket size as queried from SAI
```

Comment:
>real_bucket_size [](start = 13, length = 16)

camelCases ?

Reply:
Have changed all function names and data types to be in camelCase. Have left the underscore syntax for the struct data members as-is though. It is now consistent with how these definitions are in routeorch.

Reply:
you mix camelCases and under_score conventions for variable names in the same class.

---
In reply to: [487340692](https://github.com/Azure/sonic-swss/pull/1315#discussion_r487340692) [](ancestors = 487340692)